grype/db/v6/distribution/client.go

```diff
@@ -0,0 +1,234 @@
+	if err != nil {
+		removeAllOrLog(afero.NewOsFs(), tempDir)
+		return "", fmt.Errorf("unable to parse db URL %q: %w", c.config.LatestURL, err)
+	}
+
+	u.Path = path.Join(path.Dir(u.Path), path.Clean(archive.Path))
+
+	// from go-getter, adding a checksum as a query string will validate the payload after download
+	// note: the checksum query parameter is not sent to the server
+	query := u.Query()
+	if archive.Checksum != "" {
+		query.Add("checksum", archive.Checksum)
+	}
+	u.RawQuery = query.Encode()
+
+	// go-getter will automatically extract all files within the archive to the temp dir
+	err = c.updateDownloader.GetToDir(tempDir, u.String(), downloadProgress)
+	if err != nil {
+		removeAllOrLog(afero.NewOsFs(), tempDir)
+		return "", fmt.Errorf("unable to download db: %w", err)
+	}
+
+	return tempDir, nil
+}
+
+// latestFromURL loads a LatestDocument from a URL.
+func (c client) latestFromURL() (*LatestDocument, error) {
+	resp, err := c.latestHTTPClient.Get(c.config.LatestURL)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch latest.json: %w", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unable to fetch latest.json: %s", resp.Status)
+	}
+
+	defer resp.Body.Close()
+
+	return NewLatestFromReader(resp.Body)
+}
+
+func withClientTimeout(timeout time.Duration) func(*http.Client) {
+	return func(c *http.Client) {
+		c.Timeout = timeout
+	}
+}
+
+func defaultHTTPClient(fs afero.Fs, caCertPath string, postProcessor ...func(*http.Client)) (*http.Client, error) {
+	httpClient := cleanhttp.DefaultClient()
+	httpClient.Timeout = 30 * time.Second
+	if caCertPath != "" {
+		rootCAs := x509.NewCertPool()
+
+		pemBytes, err := afero.ReadFile(fs, caCertPath)
+		if err != nil {
+			return nil, fmt.Errorf("unable to configure root CAs for curator: %w", err)
+		}
+		rootCAs.AppendCertsFromPEM(pemBytes)
+
+		httpClient.Transport.(*http.Transport).TLSClientConfig = &tls.Config{
+			MinVersion: tls.VersionTLS12,
+			RootCAs:    rootCAs,
+		}
+	}
+
+	for _, pp := range postProcessor {
+		pp(httpClient)
+	}
+
+	return httpClient, nil
+}
+
+func removeAllOrLog(fs afero.Fs, dir string) {
+	if err := fs.RemoveAll(dir); err != nil {
+		log.WithFields("error", err).Warnf("failed to remove path %q", dir)
+	}
+}
+
+func isSupersededBy(m *v6.Description, other v6.Description) bool {
+	if m == nil {
+		log.Debugf("cannot find existing metadata, using update...")
+		// any valid update beats no database, use it!
+		return true
+	}
+
+	otherModel, otherOk := other.SchemaVersion.Model()
+	currentModel, currentOk := m.SchemaVersion.Model()
+
+	if !otherOk {
+		log.Errorf("existing database has no schema version, doing nothing...")
+		return false
+	}
+
+	if !currentOk {
+		log.Errorf("update has no schema version, doing nothing...")
+		return false
+	}
+
+	if otherModel != currentModel {
+		log.Debugf("update is for a different DB schema, skipping...")
+		return false
+	}
+
+	if other.Built.After(m.Built.Time) {
```

Comment:
note: `other.Built` is now the time when the database was _packaged_. This is a change from v5 where the same field means the time when the data was pulled, not when the data was packaged.

This change is important because if there's a flaw in the packaging, we can re-run the package step without re-running the build step and still make a new database that can be deterministically used.

We could _also_ just make grype always take the new update if the files differ at all, which would allow for server side rolling back of the database, and for always updating when switching mirrors, which is something we might discuss.